File: src/main/java/view25d/view/PatchViewOptions.java
package view25d.view;

import java.awt.Dimension;
import java.awt.GridLayout;

import javax.swing.ButtonGroup;
import javax.swing.ButtonModel;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JRadioButton;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

public class PatchViewOptions extends JPanel implements ChangeListener {

	private static final long serialVersionUID = 1L;
	PatchView myWindow;
	private JRadioButton structuresButton, sheetButton, pinsButton, threeDButton, twoDButton;


	public PatchViewOptions( PatchView parent, boolean str, boolean she, boolean pin, boolean threeD, boolean twoD ) {
		super();

		this.myWindow = parent;

		this.setLayout(new GridLayout(1,0) );
		setupUI( str, she, pin, threeD, twoD );
	}

	private void setupUI( boolean str, boolean she, boolean pin,  boolean threeD, boolean twoD  ) {

		structuresButton = new JRadioButton("Structures");
		structuresButton.setActionCommand("structures");
		if (str) { structuresButton.setSelected(true); }

		sheetButton = new JRadioButton("Sheet");
		sheetButton.setActionCommand("sheet");
		if (she) { sheetButton.setSelected(true); }

		pinsButton = new JRadioButton("Pins");
		pinsButton.setActionCommand("pins");
		if (pin) { pinsButton.setSelected(true); }

		ButtonGroup group = new ButtonGroup();
		group.add(structuresButton);
		group.add(sheetButton);
		group.add(pinsButton);

		threeDButton = new JRadioButton("xyz");
		threeDButton.setActionCommand("xyz");
		if (threeD) { threeDButton.setSelected(true); }

		twoDButton = new JRadioButton("xy-plane");
		twoDButton.setActionCommand("xy-plane");
		if (twoD) { twoDButton.setSelected(true); }

		ButtonGroup group2 = new ButtonGroup();
		group2.add(threeDButton);
		group2.add(twoDButton);

		structuresButton.addChangeListener(this);
		sheetButton.addChangeListener(this);
		pinsButton.addChangeListener(this);
		threeDButton.addChangeListener(this);
		twoDButton.addChangeListener(this);

		this.setLayout(new GridLayout(2, 4));
		this.add( new JLabel( "  Display Options:") );
		this.add( structuresButton );
		this.add( sheetButton );
		this.add( pinsButton );
		this.add( new JLabel( "  Link Options:") );
		this.add( threeDButton );
		this.add( twoDButton );
		this.setSize( new Dimension (600, 60));
	}

	public boolean sticks() { return pinsButton.isSelected(); }
	public boolean tangents() { return sheetButton.isSelected() || pinsButton.isSelected(); }
	public boolean structures() { return structuresButton.isSelected(); }
	public boolean linksAreThreeD() { return threeDButton.isSelected(); }
	public boolean linksAreTwoD() { return twoDButton.isSelected(); }
	public void setLinksDisplayMode(boolean linksInXYPlane) {
		if (linksInXYPlane) {
			twoDButton.setSelected(true);
		} else {
			threeDButton.setSelected(true);
		}

	}

	@Override
	public void stateChanged(ChangeEvent changeEvent) {

	JRadioButton aButton = (JRadioButton) changeEvent.getSource();
	ButtonModel aModel = aButton.getModel();
	String text = aButton.getText();

	if (aModel.isSelected()) {
		if (text.equals("Sheet") || text.equals("Pins") ) {
				threeDButton.setEnabled ( false );
				twoDButton.setEnabled ( false );
		}
		if (text.equals("Structures")  ) {
				threeDButton.setEnabled ( true );
				twoDButton.setEnabled ( true );
		}
	}
		myWindow.refresh();
	}
}

File: src/main/scala/view/PatchViewOptions.scala
package view25d.view

import java.awt.{ Dimension, GridBagConstraints, GridBagLayout, Insets }
import javax.swing.{ ButtonGroup, JLabel, JPanel }

import org.nlogo.swing.{ RadioButton, Transparent }
import org.nlogo.theme.{ InterfaceColors, ThemeSync }

class PatchViewOptions(window: PatchViewGUI, str: Boolean, she: Boolean, pin: Boolean, threeD: Boolean, twoD: Boolean)
  extends JPanel(new GridBagLayout) with ThemeSync {

  private val structuresButton = new RadioButton("Structures", () => {
    threeDButton.setEnabled(true)
    twoDButton.setEnabled(true)

    window.refresh()
  }) {
    if (str)
      setSelected(true)
  }

  private val sheetButton = new RadioButton("Sheet", () => {
    threeDButton.setEnabled(false)
    twoDButton.setEnabled(false)

    window.refresh()
  }) {
    if (she)
      setSelected(true)
  }

  private val pinsButton = new RadioButton("Pins", () => {
    threeDButton.setEnabled(false)
    twoDButton.setEnabled(false)

    window.refresh()
  }) {
    if (pin)
      setSelected(true)
  }

  locally {
    val group = new ButtonGroup

    group.add(structuresButton)
    group.add(sheetButton)
    group.add(pinsButton)
  }

  private val threeDButton = new RadioButton("xyz", window.refresh) {
    if (threeD)
      setSelected(true)
  }

  private val twoDButton = new RadioButton("xy-plane", window.refresh) {
    if (twoD)
      setSelected(true)
  }

  private val displayLabel = new JLabel("Display Options:")
  private val linkLabel = new JLabel("Link Options:")

  locally {
    val group = new ButtonGroup

    group.add(threeDButton)
    group.add(twoDButton)

    val c = new GridBagConstraints

    c.gridy = 0
    c.anchor = GridBagConstraints.WEST
    c.insets = new Insets(6, 6, 6, 6)

    add(displayLabel, c)

    c.insets = new Insets(6, 0, 6, 6)

    add(structuresButton, c)
    add(sheetButton, c)
    add(pinsButton, c)

    c.gridy = 1
    c.insets = new Insets(0, 6, 6, 6)

    add(linkLabel, c)

    c.insets = new Insets(0, 0, 6, 6)

    add(threeDButton, c)
    add(twoDButton, c)

    c.weightx = 1

    // invisible buffer to keep everything on the left
    add(new JPanel with Transparent, c)
  }

  setSize(new Dimension(600, 60))

  def sticks: Boolean =
    pinsButton.isSelected

  def tangents: Boolean =
    sheetButton.isSelected || pinsButton.isSelected

  def structures: Boolean =
    structuresButton.isSelected

  def linksAreThreeD: Boolean =
    threeDButton.isSelected

  def linksAreTwoD: Boolean =
    twoDButton.isSelected

  def setLinksDisplayMode(linksInXYPlane: Boolean): Unit = {
    if (linksInXYPlane)
      twoDButton.setSelected(true)
    else
      threeDButton.setSelected(true)
  }

  def syncTheme(): Unit = {
    setBackground(InterfaceColors.toolbarBackground())

    displayLabel.setForeground(InterfaceColors.toolbarText())
    linkLabel.setForeground(InterfaceColors.toolbarText())

    structuresButton.syncTheme()
    sheetButton.syncTheme()
    pinsButton.syncTheme()
    threeDButton.syncTheme()
    twoDButton.syncTheme()
  }
}
